java/vector/src/main/java/org/apache/arrow/vector/file/ArrowReader.java

```diff
@@ -17,91 +17,227 @@
 import org.apache.arrow.vector.schema.ArrowRecordBatch;
-import org.apache.arrow.vector.stream.MessageSerializer;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-public class ArrowReader implements AutoCloseable {
-  private static final Logger LOGGER = LoggerFactory.getLogger(ArrowReader.class);
-
-  public static final byte[] MAGIC = "ARROW1".getBytes();
-
-  private final SeekableByteChannel in;
-
-  private final BufferAllocator allocator;
-
-  private ArrowFooter footer;
-
-  public ArrowReader(SeekableByteChannel in, BufferAllocator allocator) {
-    super();
-    this.in = in;
-    this.allocator = allocator;
-  }
-
-  private int readFully(ByteBuffer buffer) throws IOException {
-    int total = 0;
-    int n;
-    do {
-      n = in.read(buffer);
-      total += n;
-    } while (n >= 0 && buffer.remaining() > 0);
-    buffer.flip();
-    return total;
-  }
-
-  public ArrowFooter readFooter() throws IOException {
-    if (footer == null) {
-      if (in.size() <= (MAGIC.length * 2 + 4)) {
-        throw new InvalidArrowFileException("file too small: " + in.size());
-      }
-      ByteBuffer buffer = ByteBuffer.allocate(4 + MAGIC.length);
-      long footerLengthOffset = in.size() - buffer.remaining();
-      in.position(footerLengthOffset);
-      readFully(buffer);
-      byte[] array = buffer.array();
-      if (!Arrays.equals(MAGIC, Arrays.copyOfRange(array, 4, array.length))) {
-        throw new InvalidArrowFileException("missing Magic number " + Arrays.toString(buffer.array()));
-      }
-      int footerLength = MessageSerializer.bytesToInt(array);
-      if (footerLength <= 0 || footerLength + MAGIC.length * 2 + 4 > in.size()) {
-        throw new InvalidArrowFileException("invalid footer length: " + footerLength);
-      }
-      long footerOffset = footerLengthOffset - footerLength;
-      LOGGER.debug(String.format("Footer starts at %d, length: %d", footerOffset, footerLength));
-      ByteBuffer footerBuffer = ByteBuffer.allocate(footerLength);
-      in.position(footerOffset);
-      readFully(footerBuffer);
-      Footer footerFB = Footer.getRootAsFooter(footerBuffer);
-      this.footer = new ArrowFooter(footerFB);
+import org.apache.arrow.vector.schema.VectorLayout;
+import org.apache.arrow.vector.types.Dictionary;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.Types.MinorType;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.DictionaryEncoding;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.Schema;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public abstract class ArrowReader<T extends ReadChannel> implements AutoCloseable {
+
+    private final T in;
```

Comment:
yep, the rest of the codebase is 2 spaces (as with Drill, Parquet, other ASF java codebases)